README.md
# dialog-scripts
Scripts which leverage Bart Reardon's [swiftDialog](https://github.com/bartreardon/swiftDialog).

- [Setup Your Mac](https://github.com/dan-snelson/Setup-Your-Mac/blob/main/README.md)
- [Display Message](Display%20Message/README.md)
- :new: [Disk Usage](Disk%20Usage/README.md)
- [sysdiagnose Progress](sysdiagnose%20Progress/README.md)
- [Inventory Update Progress](Inventory%20Update%20Progress/README.md)
- :new: [Notifications](swiftDialog%20Notifications/README.md)
- [swiftDialog Commander](swiftDialog%20Commander/README.md)

Continue reading at [snelson.us](https://snelson.us/) …

Setup Your Mac/CHANGELOG.md
# CHANGELOG

## 1.8.0
### 06-Mar-2023
[Release-specific Blog Post](https://snelson.us/2023/03/setup-your-mac-1-8-0-via-swiftdialog/)
- Introduces fully customizable "Configurations" (thanks, [@drtaru](https://github.com/drtaru)!)
    - **Required:** Minimum organizational apps (i.e., full disk encryption, endpoint security, VPN, communication tools, etc.)
    - **Recommended:** Required apps and Microsoft Office
    - **Complete:** Recommended apps, Adobe Acrobat Reader and Google Chrome
- Play video at Welcome dialog (Script Parameter `6`) by specifying `video` (Addresses [Issue No. 36](https://github.com/dan-snelson/dialog-scripts/issues/36))
  - 🔥 **Breaking Change** for users of Setup Your Mac prior to `1.8.0` 🔥 
      - To capture user input, `welcomeDialog` (Script Parameter `6`) must be set to `userInput`
- Addresses [Issue No. 39](https://github.com/dan-snelson/dialog-scripts/issues/39) (thanks big bunches, [@wako](https://github.com/wakco)!)
- Addresses [Issue No. 43](https://github.com/dan-snelson/dialog-scripts/issues/39) (thanks, @wako)
- Removed the following from the Welcome dialog
  - Comment
  - Select B
  - Select C

## 1.7.2
### 28-Feb-2023
- Reordered Pre-Flight Check to not validate OS until AFTER Setup Assistant / Finder & Dock
- Added `disabled` option for `requiredMinimumBuild`
- Added check for Self Service's `brandingimage.png` (Addresses [Issue No. 40](https://github.com/dan-snelson/dialog-scripts/issues/40))
- Pre-flight Check logging messages now saved to client-side log
- Addresses [Issue No. 41](https://github.com/dan-snelson/dialog-scripts/issues/41)

## 1.7.1
### 07-Feb-2023

- Addresses [Issue No. 35](https://github.com/dan-snelson/dialog-scripts/issues/35)
- Improves user-interaction with `helpmessage` under certain circumstances (thanks, @bartreardon!)
- Increased `debugMode` delay (thanks for the heads-up, @Lewis B!)
- Changed Banner Image (to something much, much smaller)

## 1.7.0
### 01-Feb-2023
[Release-specific Blog Post](https://snelson.us/2023/02/setup-your-mac-via-swiftdialog-1-7-0/)
- Adds compatibility for and leverages new features of [swiftDialog](https://github.com/bartreardon/swiftDialog/blob/main/README.md) `2.1`
  - `bannertext`
  - `helpmessage`
  - `infobox`
  - `progress`-related racing-stripes (which now **require** `swiftDialog 2.1+`)
- Completion Actions
  - Adjusted default code option (to hopefully help Mac Admins using an Enrollment Complete trigger [i.e., `runAsUser` doesn't work too well when `_mbsetupuser` is the logged-in user])
- Combined `listitem` steps for installation and validation (thanks, @roiegat!)
  - Addresses [Issue No. 30](https://github.com/dan-snelson/dialog-scripts/issues/30)
- Validate Operating System Version, Build and Outdated OS Action
  - Addresses [Issue No. 31](https://github.com/dan-snelson/dialog-scripts/issues/31)
  - Introduces the ability to specify a `requiredMinimumBuild` as Script Parameter `8` (thanks, @SirDrSpiderpig!)
    - For example, to only allow macOS Ventura 13.2 (or later), specify: `22D`
  - Introduces Outdated OS Action
    - Open Self Service to your OS update policy: `jamfselfservice://content?entity=policy&id=117&action=view`
    - Open Software Update (default): `/System/Library/CoreServices/Software Update.app`
- Temporarily disables `jamf` binary check-in (thanks, @mactroll and @cube!)
  - Purposely commented-out the code to re-enable the `jamf` binary; presumes the Mac will be restarted
- Separated "Global Variables" from "Dialog Variables" to allow for additional Script Parameters 
- Improved Pre-flight Check messaging
- Introduces `verbose` as an option for `Debug Mode`
  - Most useful when first deploying Setup Your Mac



## 1.6.0
### 09-Jan-2023
[Release-specific Blog Post](https://snelson.us/2023/01/setup-your-mac-via-swiftdialog-1-6-0/)
- Addresses [Issue No. 21](https://github.com/dan-snelson/dialog-scripts/issues/21)
  - 🔥 **Breaking Change** 🔥 (for users of Setup Your Mac prior to `1.6.0`)
      - `policy_array`'s "`path`" has been replaced with "`validation`"
  - 🆕 The `confirmPolicyExecution` function confirms if the policy needs to be executed for `filepath` validations
    - A validation of `None` always executes the Jamf Pro trigger when `debug mode` is set to `false`
  - 🆕 The `validatePolicyResult` function validates if the policy succeeded and the related service is _running_, based on the specified `validation` option
    - [Feature-specific Blog Post](https://snelson.us/2023/01/setup-your-mac-validation/)
    - **Validation Options:**
       - {absolute path} (simulates pre-`1.6.0` behavior, for example: `"/Applications/Microsoft Teams.app/Contents/Info.plist"`)
       - `Local` (for validation within this script, for example: `"filevault"`)
       - `Remote` (for validation validation via a single-script Jamf Pro policy, for example: `"symvGlobalProtect"`)
       - `None` (for triggers which don't require validation, for example: `recon`; always evaluates as successful)
- Enhanced policy logging options to address [Issue No. 25](https://github.com/dan-snelson/dialog-scripts/issues/25)
  - Search for and comment-out: `eval "${jamfBinary} policy -trigger ${trigger}"`
  - Uncomment: `eval "${jamfBinary} policy -trigger ${trigger} -verbose | tee -a ${scriptLog}"`
  - Ensure `debug mode` is set to `false`
- Added Rosetta 2 policy execution and validation
- Enhanced Logging
  - Addresses [Issue No. 29](https://github.com/dan-snelson/dialog-scripts/issues/29)


##  1.5.1
### 07-Dec-2022
[Release-specific Blog Post](https://snelson.us/2022/12/setup-your-mac-via-swiftdialog-1-5-1/)
- Updates to "Pre-flight Checks"
  - Moved section to start of script
  - :new: Added additional check for Setup Assistant (for Mac Admins using an "Enrollment Complete" trigger)

##  1.5.0
### 28-Nov-2022
[Release-specific Blog Post](https://snelson.us/2022/11/setup-your-mac-via-swiftdialog-1-5-0/)
- :new: Prompt user for additional fields at Welcome dialog

Setup Your Mac/Introduction-to-Setup-Your-Mac.bash
#!/bin/bash

####################################################################################################
#
# Introduction to Setup Your Mac via swiftDialog
# https://snelson.us/2022/12/swiftdialog-izing-your-scripts/
#
####################################################################################################
#
# HISTORY
#
#   Version 0.0.1, 03-Dec-2022, Dan K. Snelson (@dan-snelson)
#   - Greet the user immediately after Setup Assistant completes and before they start actually setting up their Mac
#
#   Version 0.0.2, 11-Dec-2022, Dan K. Snelson (@dan-snelson)
#   - Leverages swiftDialog 2.1 to play videos as logged-in user
#   - Updates to "Pre-flight Checks" from Setup Your Mac (1.5.1)
#
####################################################################################################



####################################################################################################
#
# Variables
#
####################################################################################################

scriptVersion="0.0.2"
export PATH=/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin/
scriptLog="${4:-"/var/tmp/org.churchofjesuschrist.log"}"
dialogBinary="/usr/local/bin/dialog"
dialogCommandFile=$( mktemp /var/tmp/dialogCommandFile.XXX )
loggedInUser=$( echo "show State:/Users/ConsoleUser" | scutil | awk '/Name :/ { print $3 }' )
dayOfTheWeek=$( date +'%A' )



####################################################################################################
#
# Pre-flight Checks
#
####################################################################################################

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Confirm script is running as root
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

if [[ $(id -u) -ne 0 ]]; then
    echo "This script must be run as root; exiting."
    exit 1
fi



# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Ensure computer does not go to sleep while running this script (thanks, @grahampugh!)
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

echo "Caffeinating this script (PID: $$)"
caffeinate -dimsu -w $$ &



# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Validate Setup Assistant has completed
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

while pgrep -q -x "Setup Assistant"; do
    echo "Setup Assistant is still running; pausing for 2 seconds"
    sleep 2
done

echo "Setup Assistant is no longer running; proceeding …"



# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Confirm Dock is running / user is at Desktop
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

until pgrep -q -x "Finder" && pgrep -q -x "Dock"; do
    echo "Finder & Dock are NOT running; pausing for 1 second"
    sleep 1
done

echo "Finder & Dock are running; proceeding …"



# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Validate logged-in user
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

loggedInUser=$( echo "show State:/Users/ConsoleUser" | scutil | awk '/Name :/ { print $3 }' )

if [[ -z "${loggedInUser}" || "${loggedInUser}" == "loginwindow" ]]; then
    echo "No user logged-in; exiting."
    exit 1
else
    loggedInUserFullname=$( id -F "${loggedInUser}" )
    loggedInUserFirstname=$( echo "$loggedInUserFullname" | cut -d " " -f 1 )
    # loggedInUserID=$(id -u "${loggedInUser}")
fi



####################################################################################################
#
# Introduction dialog
#
####################################################################################################

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# "Introduction" dialog Title, Message, Overlay Icon and Icon
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

title="Happy ${dayOfTheWeek}, ${loggedInUserFirstname}!"
caption="Please review this video while we get things ready."
videoID="vimeoid=774283815"



# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# "Introduction" dialog Settings and Features
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

dialogIntroduction="--title \"$title\" \
--video \"$videoID\" \
--videocaption \"$caption\" \
--infotext \"$scriptVersion\" \
--button1text \"Close\" \
--autoplay \
--moveable \
--ontop \
--width '800' \
--height '600' \
--commandfile \"$dialogCommandFile\" "



####################################################################################################
#
# Functions
#
####################################################################################################

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Client-side Script Logging
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

function updateScriptLog() {
    echo -e "$( date +%Y-%m-%d\ %H:%M:%S ) - ${1}" | tee -a "${scriptLog}"
}



# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Client-side Logging
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

if [[ ! -f "${scriptLog}" ]]; then
    touch "${scriptLog}"
    updateScriptLog "*** Created log file via script ***"
fi



# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Logging preamble
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

updateScriptLog "\n\n###\n# Introduction to Setup Your Mac (${scriptVersion})\n###\n"



# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Check for / install swiftDialog (Thanks big bunches, @acodega!)
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

function dialogCheck() {

    # Get the URL of the latest PKG From the Dialog GitHub repo
    dialogURL=$(curl --silent --fail "https://api.github.com/repos/bartreardon/swiftDialog/releases/latest" | awk -F '"' "/browser_download_url/ && /pkg\"/ { print \$4; exit }")

    # Expected Team ID of the downloaded PKG
    expectedDialogTeamID="PWA5E9TQ59"

    # Check for Dialog and install if not found
    if [ ! -e "/Library/Application Support/Dialog/Dialog.app" ]; then

        updateScriptLog "Dialog not found. Installing..."

        # Create temporary working directory
        workDirectory=$( /usr/bin/basename "$0" )
        tempDirectory=$( /usr/bin/mktemp -d "/private/tmp/$workDirectory.XXXXXX" )

        # Download the installer package
        /usr/bin/curl --location --silent "$dialogURL" -o "$tempDirectory/Dialog.pkg"

        # Verify the download
        teamID=$(/usr/sbin/spctl -a -vv -t install "$tempDirectory/Dialog.pkg" 2>&1 | awk '/origin=/ {print $NF }' | tr -d '()')

        # Install the package if Team ID validates
        if [[ "$expectedDialogTeamID" == "$teamID" ]]; then

            /usr/sbin/installer -pkg "$tempDirectory/Dialog.pkg" -target /
            sleep 2
            updateScriptLog "swiftDialog version $(dialog --version) installed; proceeding..."

        else

            # Display a so-called "simple" dialog if Team ID fails to validate
            runAsUser osascript -e 'display dialog "Please advise your Support Representative of the following error:\r\r• Dialog Team ID verification failed\r\r" with title "Setup Your Mac: Error" buttons {"Close"} with icon caution'
            quitScript "1"

        fi

        # Remove the temporary working directory when done
        /bin/rm -Rf "$tempDirectory"  

    else

        updateScriptLog "swiftDialog version $(dialog --version) found; proceeding..."

    fi

}



# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Update the "Introduction" dialog
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

function dialogUpdateIntroduction(){
    updateScriptLog "INTRODUCTION DIALOG: $1"
    echo "$1" >> "$dialogCommandFile"
    sleep 0.35
}



# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Kill a specified process (thanks, @grahampugh!)
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

function killProcess() {
    process="$1"
    if process_pid=$( pgrep -a "${process}" 2>/dev/null ) ; then
        updateScriptLog "Attempting to terminate the '$process' process …"
        updateScriptLog "(Termination message indicates success.)"
        kill "$process_pid" 2> /dev/null
        if pgrep -a "$process" >/dev/null ; then
            updateScriptLog "ERROR: '$process' could not be terminated."
        fi
    else
        updateScriptLog "The '$process' process isn't running."
    fi
}



# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Quit Script (thanks, @bartreadon!)
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

function quitScript() {

    updateScriptLog "Quitting …"
    dialogUpdateIntroduction "quit:"

    # Stop `caffeinate` process
    updateScriptLog "De-caffeinate …"
    killProcess "caffeinate"

    # Remove dialogCommandFile
    if [[ -e ${dialogCommandFile} ]]; then
        updateScriptLog "Removing ${dialogCommandFile} …"
        rm "${dialogCommandFile}"
    fi

    exit "${1}"

}



####################################################################################################
#
# Program
#
####################################################################################################

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Validate swiftDialog is installed
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

dialogCheck



# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Display Introduction dialog
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

eval "${dialogBinary} --args ${dialogIntroduction}"
# su - "${loggedInUser}" -c "open -a ${dialogBinary} --args ${dialogIntroduction} & sleep 0.35"



# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# Exit
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

quitScript "0"